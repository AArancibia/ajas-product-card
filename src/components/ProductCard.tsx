import React from "react";
import { createContext, CSSProperties } from "react";
import styles from '../styles/styles.module.css';

import { useProduct } from "../hooks/useProduct";
import {ProductContextProps,Product, onChangeArgs, InitialValues, ProductCardHandlers} from '../interfaces/interfaces';


export const ProductContext = createContext({} as ProductContextProps);
const {Provider} = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({children, product, value, className, style, onChange, initialValues}: Props) => {
    
    const {counter, increaseBy, maxCount, isMaxCountReached, reset} = useProduct({onChange, product, value, initialValues});

    return (
      <Provider value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}>
        <div 
          className={`${styles.productCard} ${className}`}
          style={style}
        >
          {
            children({
              count: counter,
              isMaxCountReached,
              maxCount: initialValues?.count,
              increaseBy,
              reset,
              product,
            })
          }
        </div>
      </Provider>
    );
}


