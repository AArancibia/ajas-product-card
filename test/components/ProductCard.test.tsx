import React from "react";
import { ProductCard } from "../../src/components";
import renderer, { act } from "react-test-renderer";
import { product1 } from "../data/products";

describe('ProductImage', () => {

    test('should show the component with a correct title', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <h1>Hola</h1>
                    )
                }
            </ProductCard>
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
      });

      test('should show the component with a correct title', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    ({count, increaseBy}) => (
                        <>
                            <h1>Hola</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}></button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
        
        act(() => {
            (tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();
        
        expect((tree as any).children[1].children[0]).toMatchSnapshot();
      });

      
})