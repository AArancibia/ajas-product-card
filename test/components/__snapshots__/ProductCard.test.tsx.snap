// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show the component with a correct title 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hola
  </h1>
</div>
`;

exports[`ProductImage should show the component with a correct title 2`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hola
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`ProductImage should show the component with a correct title 3`] = `"1"`;
